feat: 공통 모달 컴포넌트 구현 (#29)

yeeun-kor opened the pull request:
## 📑 구현 사항
### 1. 모달 여는 방식이 2가지로 구현
- **trigger 버튼**: props로 버튼 텍스트, 아이콘, variant 설정해야함
- **이중모달 제어**: open/onOpenChange로 외부 state 제어 (2중 모달 지원)

### 2. 재사용 가능한 구조
- `content` props로 어떤 컴포넌트든 전달 가능 (폼, 카드, 그리드 등)
- `title`, `description` props 로 헤더 커스터마이징하여 개발자가 입력 가능
- Separator로 섹션 구분했음

### 3. 옵셔널 Footer
- `footer` props가 없으면 Footer 영역 미표시
- `footer.description`: footer의 좌측 설명 텍스트
- `footer.closeButton`, `footer.submitButton`: 버튼 커스터마이징
  - 텍스트, 아이콘, variant 설정 가능
  -  버튼 기본값은 `variant: closeButton='outline', submitButton='primary'`

## 🌀 어려웠던 점 / 고민했던 부분
### 2중 모달의 구조
#### 고민했던 부분
- 첫 번째 모달의 content 안에서 두 번째 모달을 여는 구조를 구현해야 했습니다.
- 그러기 위해 상태 값으로 모달창을 열고 닫게 하는 방법으로 구현을 시도했습니다

#### open={isDetailModalOpen}
- 현재 모달이 열려있는지 여부를 나타내는 상태값(just on/off)
- true → 모달 열림
- false → 모달 닫힘

#### onOpenChange={setIsDetailModalOpen}
- 모달의 열림/닫힘 상태가 변경될 때 호출되는 함수
- 사용자가 X 버튼 클릭or배경 클릭orESC 키 누르면 → onOpenChange(false) 호출됨
- 그럼 setIsDetailModalOpen(false)가 실행되어 모달이 닫힘

> 굳이 상태값이 2개인 이유는? 닫는 버튼이 하나인게 아니고, 여러 상황(버튼과 배경클릭과 x버튼) 에서도 모달이 닫혀져야함.그래서 상태값을 2개로 관리해야함.


### Separator의 전체 너비 확장

#### 어려웠던 점
- 라이브러리가 제공해주는 `DialogContent`에 padding: 24px이 있어서, `Separator`가 padding 안쪽에만 그려졌습니다.
- 디자인상 Separator는 모달 전체 너비를 차지해야 했습니다.

#### 시도한 방법:
- Tailwind의 음수 마진 사용 → 반응형 대응 어렵고 전체 너비는 동일했음 
- 인라인 스타일로 계산된 값 사용하여 전체너비도 부모 너비를 반영하도록 설계함
#### 현재 코드:
```tsx
<Separator
  style={{
    marginLeft: '-24px',
    marginRight: '-24px',
    width: 'calc(100% + 48px)',
    maxWidth: 'none',
  }}
/>
```


## 📸 구현 이미지 (선택)
![모달](https://github.com/user-attachments/assets/4f1595ef-5efb-48c6-8543-95a334b2c13f)


## ❇️ 코드 설명 (선택)
### 사용 예시 1: trigger 버튼으로 열기
```tsx
<Modal
  trigger={{ text: '지원하기', variant: 'primary', icon: <Send /> }}
  title="스터디 지원"
  description="지원서 작성"
  content={<FormComponent />}
  footer={{
    description: '*필수 항목입니다',
    closeButton: { text: '취소' },
    submitButton: { text: '지원하기', icon: <Send /> },
  }}
/>
```
### 예시 2: 프로그래밍 방식 (2중 모달)
```tsx
const [isOpen, setIsOpen] = useState(false)

// 첫 번째 모달
<Modal
  trigger={{ text: '지원내역' }}
  content={
    <div onClick={() => setIsOpen(true)}>
      지원자 카드 리스트
    </div>
  }
/>

// 두 번째 모달 (첫 번째 모달 안에서 열림)
{isOpen && (
  <Modal
    open={isOpen}
    onOpenChange={setIsOpen}
    title="지원자 상세 정보"
    content={<DetailComponent />}
    footer={{
      closeButton: { text: '거절', variant: 'danger' },
      submitButton: { text: '승인', variant: 'success' },
    }}
  />
)}
```

### 예시 3: Footer 없는 모달
```tsx
<Modal
  trigger={{ text: '공지사항' }}
  title="공지사항"
  content={<NoticeComponent />}
  // footer 없음
/>
```

## 💬 리뷰 요청사항 (선택)
1. 2중 모달 구현 방식 :
-  현재 open/onOpenChange props로 2중 모달을 구현했습니다. 
- 더 나은 패턴이나 Radix UI의 권장 방식이 있는지 의견 부탁드립니다.
2. Form 태그 사용 
- <form> 태그로 모든 모달을 감싸고 있는데, 모든 모달이 form을 사용하지 않을 수도 있잖아요?
-  이 부분을 props로 제어하는 게 좋을까요? 아니면 외부에서 content에 form을 넣는 방식이 나을까요?

## 💬 리뷰요청에 대한 코드 리팩토링 시행

### 리팩토링 배경
초기 구현에서는 footer의 버튼을 `closeButton`, `submitButton` 객체로 받아 
Modal 내부에서 Button을 렌더링하는 구조였습니다.

하지만 지훈님과의 코드 리뷰를 통해 다음과 같은 문제점을 발견했습니다:
1. **유연성 부족**: 버튼 개수가 2개로 고정됨 (거절/승인처럼 여러 버튼 필요 시 대응 불가)
2. **제어 제한**: `type`, `onClick` 등 Button의 세부 속성 제어 불가
3. **학습 비용**: Modal의 footer 구조를 별도로 학습해야 함

### 리팩토링 내용
`footer.submitButton` 객체를 제거하고, `footer.footerButtons`로 **Button 컴포넌트 자체**를 전달받도록 변경했습니다.

#### 변경 사항
- **삭제**: `submitButton: { text, variant, icon }` 객체 구조
- **추가**: `footerButtons: React.ReactNode` - <u>사용자가 직접 Button 컴포넌트 전달</u>
- **유지**: `closeButton: { text }` - DialogClose로 감싸야 하므로 유지


#### 리팩토링 전 코드 
```tsx
<Modal
  trigger={{ text: '지원하기', variant: 'primary', icon: <Send /> }}
  title="스터디 지원"
  description="지원서 작성"
  content={<FormComponent />}
  footer={{
    description: '*필수 항목입니다',
    closeButton: { text: '취소' },
    submitButton: { text: '지원하기', icon: <Send /> },
  }}
/>
```
#### 리팩토링 후 코드 
```tsx
      <Modal
        trigger={{ text: '지원하기', variant: 'primary', icon: <Send /> }}
        title="스터디 지원"
        description="지원서 작성"
        content={<div>컨텐츠내용</div>}
        footer={{
          description: '*필수 항목입니다',
          closeButton: { text: '취소' },
          footerButtons: (
            <Button type="submit">
              <Send /> 지원하기
            </Button>
          ),
        }}
      />
```

### 리팩토링 효과
#### 1. 여러 버튼 지원 가능
```tsx
// 승인/거절 버튼을 동시에 표시
footer={{
  footerButtons: (
    <>
      <Button variant="danger">거절</Button>
      <Button variant="success">승인</Button>
    </>
  ),
}}
```
#### 2. type 속성 제어 가능
```tsx
// submit이 아닌 button 타입도 가능
<Button type="button" onClick={handleCustomAction}>
  커스텀 액션
</Button>
```

## 결론
공통 컴포넌트는 제약이 적고 유연할수록 좋다는 원칙에 따라, 사용자에게 더 많은 제어권을 주는 방향으로 리팩토링했습니다. 
처음 개발 당시 더 많은 제어권을 주게되면, `공통컴포넌트`로써 역할이 뭘까? 
더 지저분하고 보일러플레이트 코드가 하나 더 생기는거 아닌가? 싶었습니다. 
그러나, 버튼태그 한줄 더 쓰는것이 footer-props구조를 익히는 것보다 더 쉽다는걸 깨닫고 
이를 통해 Modal을 사용하는 개발자 입장에서 학습 비용은 줄이고, 활용 가능성은 높이는 구조를 만들 수 있었습니다.

agrade1 committed:
feat: header 알람 모달 연동
- header의 알람 버튼 클릭히 모달 오픈

agrade1 committed:
feat: alarm 반응형 추가
- alarm 모달 반응형 작업 추가

yeeun-kor committed:
chore: shadcn 모달패키지 설치
- npx shadcn@latest add dialog

yeeun-kor committed:
chore: shadcn 모달패키지 설치
- npx shadcn@latest add dialog

agrade1 committed:
feat: alarm 목록 조회 api 작업 및 연동
- alarm 목록 조회 api 작업
- 받아온 데이터를 가공하는 mapper 구현
- 매핑된 데이터로 alarmCard 노출

yeeun-kor committed:
chore: shadcn 모달창에 필요한 라이브러리 추가 설치
- input, label 라이브러리 설치

yeeun-kor committed:
Merge branch 'feat/29-dialog-components' of https://github.com/OZ-Coding-School/oz_externship_fe_04_team3 into feat/29-dialog-components

agrade1 committed:
feat: noti 필터링 api 연결
- 필터링 버튼 클릭시 쿼리 값 설정해서 api 요청

agrade1 committed:
feat: 알림 읽기, 전체 읽기 api 연동
- 알림 개별 읽기, 전체 읽기 mock api 구현
- 전체 읽기 후 목록 재조회
- 개별 읽기 후 리다이렉트 골조 구현(추후 페이지 작업 완료될시 실제 url로 이동하게끔 구현 필요 사항)

agrade1 committed:
chore: radix-label 패키지 설치
- radix-ui/react-label 패키지 설치 누락으로 설치 진행

yeeun-kor committed:
chore: npm패키지 설치
- label 누락되어서 설치함
- separator 구분선 라이브러리 추가

yeeun-kor committed:
chore: npm패키지 설치
- label 누락되어서 설치함
- separator 구분선 라이브러리 추가

yeeun-kor committed:
feat: 스타일 수정
- select 컴포넌트 공백제거
- 라벨 및 구분선 추가

yeeun-kor committed:
feat: 스타일 수정
- select 컴포넌트 공백제거
- 라벨 및 구분선 추가

yeeun-kor committed:
Merge branch 'develop' of https://github.com/OZ-Coding-School/oz_externship_fe_04_team3 into feat/29-dialog-components

yeeun-kor committed:
feat: 공통 Modal 컴포넌트 구현
- Radix UI Dialog 기반 재사용 가능한 컴포넌트 추가
- trigger 버튼 / 프로그래밍 제어 방식 지원
- content, footer 커스터마이징 가능
- 2중 모달 지원

yeeun-kor committed:
feat: 공통 Modal 컴포넌트 구현
- Radix UI Dialog 기반 재사용 가능한 컴포넌트 추가
- trigger 버튼 / 프로그래밍 제어 방식 지원
- content, footer 커스터마이징 가능
- 2중 모달 지원

agrade1 reviewed (dismissed):
코드 설명이 깔끔해서 추가 질문이 없내여 고생하셨습니다!

KoCeleste reviewed (dismissed):
고민하신 흔적이 많아보이고 Separator 관련 부분 정리해주셔서 보기 좋은 것 같습니다.

yeeun-kor committed:
Merge branch 'feat/29-dialog-components' of https://github.com/OZ-Coding-School/oz_externship_fe_04_team3 into feat/29-dialog-components

yeeun-kor committed:
refactor: Modal footer 버튼 구조 개선
- 버튼은 props가 아닌 직접 전달 할 것
- content의 만약 form태그를 사용하는 컴포넌트라면 Button의 form속성과 연결할 수 있는 장점도 갖고 있음.
- 코드 리뷰 피드백 반영 (지훈님 제안)

agrade1 reviewed (approved):
리팩토링도 잘 적용된거같습니다! :)

KoCeleste reviewed (approved):
리팩토링까지 수고많으셨습니다👍
